### ISSUE TITLE
Fatal error. System.AccessViolationException

### ISSUE THREAD

**Issue by paultobey** [xunit/xunit#2503]
### Discussed in https://github.com/xunit/xunit/discussions/2501

<div type='discussions-op-text'>

<sup>Originally posted by **paultobey** March 29, 2022</sup>
xUnit 2.4.1

I'm getting intermittent crashes of a large, multiple test assembly, test set (1800+ tests, 4 test assemblies, 150 tests creating/using/dropping SQL Server databases):

```
dotnet test Core.sln --framework netcoreapp3.1 --no-build --no-restore --filter TestCategory!=Performance --verbosity normal >test2.out
The active test run was aborted. Reason: Test host process crashed : Fatal error. System.AccessViolationException: Attempted to read or write protected memory. This is often an indication that other memory is corrupt.
   at System.Diagnostics.StackTrace.GetStackFramesInternal(System.Diagnostics.StackFrameHelper, Int32, Boolean, System.Exception)
   at System.Diagnostics.StackFrameHelper.InitializeSourceInfo(Int32, Boolean, System.Exception)
   at System.Diagnostics.StackTrace.CaptureStackTrace(Int32, Boolean, System.Exception)
   at System.Diagnostics.StackTrace..ctor(System.Exception, Boolean)
   at System.Exception.get_StackTrace()
   at Xunit.Sdk.ExceptionUtility.ConvertExceptionToFailureInformation(System.Exception, Int32, System.Collections.Generic.List`1<System.String>, System.Collections.Generic.List`1<System.String>, System.Collections.Generic.List`1<System.String>, System.Collections.Generic.List`1<Int32>)
   at Xunit.Sdk.ExceptionUtility.ConvertExceptionToFailureInformation(System.Exception)
   at Xunit.Sdk.TestFailed..ctor(Xunit.Abstractions.ITest, System.Decimal, System.String, System.Exception)
   at Xunit.Sdk.TestRunner`1+<RunAsync>d__43[[System.__Canon, System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].MoveNext()
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1+AsyncStateMachineBox`1[[System.__Canon, System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e],[Xunit.Sdk.TestRunner`1+<RunAsync>d__43[[System.__Canon, System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]], xunit.execution.dotnet, Version=2.4.1.0, Culture=neutral, PublicKeyToken=8d05b1bb7a6fdb6c]].ExecutionContextCallback(System.Object)
   at System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1+AsyncStateMachineBox`1[[System.__Canon, System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e],[Xunit.Sdk.TestRunner`1+<RunAsync>d__43[[System.__Canon, System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]], xunit.execution.dotnet, Version=2.4.1.0, Culture=neutral, PublicKeyToken=8d05b1bb7a6fdb6c]].MoveNext(System.Threading.Thread)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1+AsyncStateMachineBox`1[[System.__Canon, System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e],[Xunit.Sdk.TestRunner`1+<RunAsync>d__43[[System.__Canon, System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]], xunit.execution.dotnet, Version=2.4.1.0, Culture=neutral, PublicKeyToken=8d05b1bb7a6fdb6c]].MoveNext()
   at System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(System.Runtime.CompilerServices.IAsyncStateMachineBox, Boolean)
   at System.Threading.Tasks.Task.RunContinuations(System.Object)
   at System.Threading.Tasks.Task`1[[System.__Canon, System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].TrySetResult(System.__Canon)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1[[System.__Canon, System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].SetResult(System.__Canon)
   at Xunit.Sdk.ExceptionAggregator+<RunAsync>d__10`1[[System.__Canon, System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].MoveNext()
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1+AsyncStateMachineBox`1[[System.__Canon, System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e],[Xunit.Sdk.ExceptionAggregator+<RunAsync>d__10`1[[System.__Canon, System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]], xunit.core, Version=2.4.1.0, Culture=neutral, PublicKeyToken=8d05b1bb7a6fdb6c]].ExecutionContextCallback(System.Object)
   at System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1+AsyncStateMachineBox`1[[System.__Canon, System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e],[Xunit.Sdk.ExceptionAggregator+<RunAsync>d__10`1[[System.__Canon, System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]], xunit.core, Version=2.4.1.0, Culture=neutral, PublicKeyToken=8d05b1bb7a6fdb6c]].MoveNext(System.Threading.Thread)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1+AsyncStateMachineBox`1[[System.__Canon, System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e],[Xunit.Sdk.ExceptionAggregator+<RunAsync>d__10`1[[System.__Canon, System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]], xunit.core, Version=2.4.1.0, Culture=neutral, PublicKeyToken=8d05b1bb7a6fdb6c]].MoveNext()
   at System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(System.Runtime.CompilerServices.IAsyncStateMachineBox, Boolean)
   at System.Threading.Tasks.Task.RunContinuations(System.Object)
   at System.Threading.Tasks.Task`1[[System.__Canon, System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].TrySetResult(System.__Canon)
   at Xunit.Sdk.XunitTestRunner+<InvokeTestAsync>d__4.MoveNext()
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1+AsyncStateMachineBox`1[[System.__Canon, System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e],[Xunit.Sdk.XunitTestRunner+<InvokeTestAsync>d__4, xunit.execution.dotnet, Version=2.4.1.0, Culture=neutral, PublicKeyToken=8d05b1bb7a6fdb6c]].ExecutionContextCallback(System.Object)
   at System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1+AsyncStateMachineBox`1[[System.__Canon, System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e],[Xunit.Sdk.XunitTestRunner+<InvokeTestAsync>d__4, xunit.execution.dotnet, Version=2.4.1.0, Culture=neutral, PublicKeyToken=8d05b1bb7a6fdb6c]].MoveNext(System.Threading.Thread)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1+AsyncStateMachineBox`1[[System.__Canon, System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e],[Xunit.Sdk.XunitTestRunner+<InvokeTestAsync>d__4, xunit.execution.dotnet, Version=2.4.1.0, Culture=neutral, PublicKeyToken=8d05b1bb7a6fdb6c]].MoveNext()
   at System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(System.Runtime.CompilerServices.IAsyncStateMachineBox, Boolean)
   at System.Threading.Tasks.Task.RunContinuations(System.Object)
   at System.Threading.Tasks.Task`1[[System.Decimal, System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].TrySetResult(System.Decimal)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1[[System.Decimal, System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].SetExistingTaskResult(System.Decimal)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1[[System.Decimal, System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].SetResult(System.Decimal)
   at Xunit.Sdk.ExceptionAggregator+<RunAsync>d__10`1[[System.Decimal, System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].MoveNext()
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1+AsyncStateMachineBox`1[[System.Decimal, System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e],[Xunit.Sdk.ExceptionAggregator+<RunAsync>d__10`1[[System.Decimal, System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]], xunit.core, Version=2.4.1.0, Culture=neutral, PublicKeyToken=8d05b1bb7a6fdb6c]].ExecutionContextCallback(System.Object)
   at System.Threading.ExecutionContex
Test Run Aborted.
```

The point in the process at which the test crashes varies and there does not appear to be a correlation with any individual tests which are in process. Any command-line switches that would deliver better information? Any other suggestions? I'll try to set up memory dumps when it fails...

Thanks,
Paul T.

-----

Update 1:

Here's a sample dump file when the error occurs:

[dotnet.exe.34148.dmp.zip](https://github.com/xunit/xunit/files/8373973/dotnet.exe.34148.dmp.zip)

-----

Creating an issue from the unanswered discussion. Sorry I can't provide a simple replication case. This happens 1/4 of the time when running 'dot net test' as noted above.

Thanks,
Paul T.

</div>

**Comment by bartelink** [comment]
https://andrewlock.net/tracking-down-a-hanging-xunit-test-in-ci-building-a-custom-test-framework/
See also https://twitter.com/andrewlocknet/status/1508845968833363973
I really don't get the point in you logging this without a repro
Unfortunately, the chances are SELECT is not broken, and you have debugging to do
Perhaps at the end of narrowing things down, you might be able to formulate a feature idea that would help you e.g. an ordered list of which tests ran and/or how they overlapped. But, even then, whether that's directly a XUnit feature will be debatable

**Comment by bartelink** [comment]
The problem is that GH discussions and issues are both low traffic and rely on two things:
- people seeing them
- people getting enough from them to be able to answer them at all

SO is better in terms of how many will see them - only people that watch the repo will see these issues,

While many people around the planet can, given time and a dump, infer things, the chances are they wont  as your work so far has yield the universe the following contextual info:

>xUnit 2.4.1

>I'm getting intermittent crashes of a large, multiple test assembly, test set (1800+ tests, 4 test assemblies, 150 tests creating/using/dropping SQL Server databases):

> The point in the process at which the test crashes varies and there does not appear to be a correlation with any individual tests which are in process. Any command-line switches that would deliver better information? Any other suggestions? I'll try to set up memory dumps when it fails...

TL;DR I'd hate you to think either
1. its likely that this is down to a bug in Xunit
2. you've provided useful information in a good place so that someone will eventually come along and debug this

All there is here now is:
1. something that a maintainer cannot reasonably do anything useful with
2. noise that watchers like encounter

In general for AccessViolation exceptions, its a matter of narrowing down different libs you hit in the context of your tests and/or things you trigger by doign things concurrently.

I'd be doing this like:
- constrain parallelism to avoid concurrency being a factor
- try to exclude groups of tests from runs etc

The stuff I linked to can give circumstantial evidence about test order etc, but my key tip would be: the test framework is definitely the last suspect, even if everything you can rule out is a win in the process of narrowing things down.

I've sat in your seat debugging stuff like this and its definitely no fun though, as there's no real roadmap.... So all I can do is wish you luck in your debugging.

**Comment by attentive** [comment]
I was getting this error message recently.

I guessed in my case that it was due to parallelised tests contending over shared data structures in linked native libraries (in this case one or more underlying format drivers that go along with the GDAL.Native NuGet package).

The reason for the hunch was that I know some of the underlying native libraries are not properly threadsafe (The C/C++ codebases for them are 20+ years old in some cases). The underlying code was also opening file handles on the same source test datasets across multiple test collections with slightly tweaked fixtures, and so on.

Anyway, I couldn't figure out a way to get the tests to run in parallel, but overriding the default `CollectionBehavior` immediately stopped the access violation errors.

```cs
global using Xunit;

[assembly: CollectionBehavior(DisableTestParallelization = true)]
```

I hope this comment will be helpful for someone else who drops past this thread experiencing one of these glitches.

**Comment by bradwilson** [comment]
@attentive There are two issues here.

First, xUnit.net should not be hard crashing here. That's something that should be fixable.

Second, it appears that some library is corrupting memory. Catching this exception so that it doesn't hard crash isn't guaranteed to fix everything, since corrupted memory issues may then spring up all over, depending on the source and extent of the corruption. So it may not necessarily be super helpful for us (in this case) to just return back the original error with no stack trace attached, since it's the parsing of the stack trace that appears to be causing the problem. My gut feeling is that a corrupted stack trace is _very_ bad.

**Comment by bradwilson** [comment]
(Note: there's a copy/paste error in the v2 tree that I didn't catch until after I pushed: https://github.com/xunit/xunit/commit/57af1d9f0b18181631cf6aa3d40a9f950468b1bc)

**Comment by bradwilson** [comment]
Attempted fix available in v2: `2.7.1-pre.14`, v3: `0.1.1-pre.391`

https://xunit.net/docs/using-ci-builds

**Comment by attentive** [comment]
Very responsive @bradwilson 👏  … I have made a note to see how this goes soon (might be after Easter).

I didn't really consider this an xUnit bug though it was a bit confusing at first, but also I knew I was doing some pretty hectic stuff with some pretty old underlying libraries.
